File: Example/ObjCExample/Core Data/SavePhotosOperation.m
//
//  SavePhotosOperation.m
//  ASJCoreDataOperationExample
//
//  Created by sudeep_MAC02 on 12/05/16.
//  Copyright © 2016 sudeep. All rights reserved.
//

#import "SavePhotosOperation.h"
#import "Photo.h"

@implementation SavePhotosOperation

- (void)coreDataOperation
{
    for (NSDictionary *photoInfo in _photos)
    {
        NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] initWithEntityName:@"Photo"];
        fetchRequest.fetchLimit = 1;
        
        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"photoId == %@", photoInfo[@"id"]];
        fetchRequest.predicate = predicate;
        
        NSError *error = nil;
        NSArray *result = [self.privateMoc executeFetchRequest:fetchRequest error:&error];
        if (error) {
            NSLog(@"error fetching existing photo: %@", error.localizedDescription);
        }
        
        if (result.count > 0) {
            continue;
        }
        
        Photo *photoManagedObject = (Photo *)[NSEntityDescription insertNewObjectForEntityForName:@"Photo" inManagedObjectContext:self.privateMoc];
        
        photoManagedObject.albumId = photoInfo[@"albumId"];
        photoManagedObject.photoId = photoInfo[@"id"];
        photoManagedObject.title = photoInfo[@"title"];
        photoManagedObject.url = photoInfo[@"url"];
        photoManagedObject.thumbnailUrl = photoInfo[@"thumbnailUrl"];
        
        BOOL success = [self.privateMoc save:&error];
        if (!success || error) {
            NSLog(@"error saving photo: %@", error.localizedDescription);
        }
    }
}

@end

File: Examples/Swift/SwiftExample/Core Data/SavePhotosOperation.swift
//
//  SavePhotosOperation.swift
//  CoreDataOperation
//
//  Created by sudeep on 23/05/16.
//  Copyright © 2016 sudeep. All rights reserved.
//

import UIKit
import ASJCoreDataOperation_Swift
import CoreData

public class SavePhotosOperation: ASJCoreDataOperation
{
    public var photos: [[String: AnyObject]] = []
    
    override public func coreDataOperation()
    {
        for photoInfo in photos
        {
            let fetchRequest = NSFetchRequest<Photo>(entityName: "Photo")
            fetchRequest.fetchLimit = 1
            
            if let photoId = photoInfo["id"] as? NSNumber
            {
                let predicate = NSPredicate(format: "photoId == %@", photoId)
                fetchRequest.predicate = predicate
            }
            
            do
            {
                let result = try privateMoc.fetch(fetchRequest) as [Photo]
                if result.count > 0 {
                    continue
                }
            }
            catch let error as NSError {
                print("error fetching existing photos: \(error.localizedDescription)")
            }
            
            let photoManagedObject = NSEntityDescription.insertNewObject(forEntityName: "Photo", into: privateMoc) as! Photo
            
            photoManagedObject.albumId = photoInfo["albumId"] as? NSNumber
            photoManagedObject.photoId = photoInfo["id"] as? NSNumber
            photoManagedObject.title = photoInfo["title"] as? String
            photoManagedObject.url = photoInfo["url"] as? String
            photoManagedObject.thumbnailUrl = photoInfo["thumbnailUrl"] as? String
            
            do {
                try privateMoc.save()
            }
            catch let error as NSError {
                print("error saving photo: \(error.localizedDescription)")
            }
        }
        
    }
}
